src/ui/public/directives/match_pairs.js

```diff
@@ -0,0 +1,101 @@
+import { uiModules } from 'ui/modules';
+const module = uiModules.get('kibana');
+
+/**
+ * This directively automatically handles matching pairs.
+ * Specifically, it does the following:
+ *
+ * 1. If the key is a closer, and the character in front of the cursor is the
+ *    same, simply move the cursor forward.
+ * 2. If the key is an opener, insert the opener at the beginning of the
+ *    selection, and the closer at the end of the selection, and move the
+ *    selection forward.
+ * 3. If the backspace is hit, and the characters before and after the cursor
+ *    are a pair, remove both characters and move the cursor backward.
+ */
+
+const pairs = ['()', '[]', '{}', `''`, '""'];
+const openers = pairs.map(pair => pair[0]);
+const closers = pairs.map(pair => pair[1]);
+
+module.directive('matchPairs', () => ({
+  restrict: 'A',
+  require: 'ngModel',
+  link: function (scope, elem, attrs, ngModel) {
+    elem.on('keydown', (e) => {
+      const { target, key, metaKey } = e;
+      const { value, selectionStart, selectionEnd } = target;
+
+      if (shouldMoveSelectionForward()) {
+        moveSelectionForward();
+      } else if (shouldInsertMatchingCloser()) {
+        insertMatchingCloser();
+      } else if (shouldRemovePair()) {
+        removePair();
+      }
+
+      function shouldMoveSelectionForward() {
+        if (!closers.includes(key)) return false;
+
+        // Never move selection forward for multi-character selections
+        if (selectionStart !== selectionEnd) return false;
+
+        // Move selection foward if the key is the same as the closer in front of the selection
+        return value.charAt(selectionEnd) === key;
+      }
+
+      function shouldInsertMatchingCloser() {
+        if (!openers.includes(key)) return false;
+
+        // Always insert for multi-character selections
+        if (selectionStart !== selectionEnd) return true;
+
+        const precedingCharacter = value.charAt(selectionStart - 1);
+        const followingCharacter = value.charAt(selectionStart + 1);
+
+        // Don't insert if the preceding character is a backslash
+        if (precedingCharacter === '\\') return false;
```

Comment:
This makes me think this module is somewhat coupled to our KQL syntax. Should we rename and move the file to make that more apparent since it's not totally generic?